bemani/backend/iidx/heroicverse.py

```diff
@@ -0,0 +1,1883 @@
+            [
+                qpro_dict.get_int('head'),
+                qpro_dict.get_int('hair'),
+                qpro_dict.get_int('face'),
+                qpro_dict.get_int('hand'),
+                qpro_dict.get_int('body'),
+            ],
+        ))
+
+        # Rivals
+        rlist = Node.void('rlist')
+        root.add_child(rlist)
+        links = self.data.local.user.get_links(self.game, self.version, userid)
+        for link in links:
+            rival_type = None
+            if link.type == 'sp_rival':
+                rival_type = '1'
+            elif link.type == 'dp_rival':
+                rival_type = '2'
+            else:
+                # No business with this link type
+                continue
+
+            other_profile = self.get_profile(link.other_userid)
+            if other_profile is None:
+                continue
+            other_play_stats = self.get_play_statistics(link.other_userid)
+
+            rival = Node.void('rival')
+            rlist.add_child(rival)
+            rival.set_attribute('spdp', rival_type)
+            rival.set_attribute('id', str(other_profile.get_int('extid')))
+            rival.set_attribute('id_str', ID.format_extid(other_profile.get_int('extid')))
+            rival.set_attribute('djname', other_profile.get_str('name'))
+            rival.set_attribute('pid', str(other_profile.get_int('pid')))
+            rival.set_attribute('sg', str(self.db_to_game_rank(other_profile.get_int(self.DAN_RANKING_SINGLE, -1), self.GAME_CLTYPE_SINGLE)))
+            rival.set_attribute('dg', str(self.db_to_game_rank(other_profile.get_int(self.DAN_RANKING_DOUBLE, -1), self.GAME_CLTYPE_DOUBLE)))
+            rival.set_attribute('sa', str(other_play_stats.get_int('single_dj_points')))
+            rival.set_attribute('da', str(other_play_stats.get_int('double_dj_points')))
+
+            rival.add_child(Node.bool('is_robo', False))
+
+            # If the user joined a particular shop, let the game know.
+            if 'shop_location' in other_profile:
+                shop_id = other_profile.get_int('shop_location')
+                machine = self.get_machine_by_id(shop_id)
+                if machine is not None:
+                    shop = Node.void('shop')
+                    rival.add_child(shop)
+                    shop.set_attribute('name', machine.name)
+
+            qprodata = Node.void('qprodata')
+            rival.add_child(qprodata)
+            qpro = other_profile.get_dict('qpro')
+            qprodata.set_attribute('head', str(qpro.get_int('head')))
+            qprodata.set_attribute('hair', str(qpro.get_int('hair')))
+            qprodata.set_attribute('face', str(qpro.get_int('face')))
+            qprodata.set_attribute('body', str(qpro.get_int('body')))
+            qprodata.set_attribute('hand', str(qpro.get_int('hand')))
+
+            challenge = Node.void('challenge')
+            rival.add_child(challenge)
+
+        # Expert  (removed in rootage)
```

Comment:
Things removed outright can just be deleted. I leave commented out code where somebody coming by to implement a feature I missed will have a starting point.